### PR TITLE
ocamlPackages.opium: init at 0.17.1

### PR DESCRIPTION
**Pull request by pmahoney**
###### Motivation for this change

To use opium ocaml library (and hmap, a dependency of opium)

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nix-review --run "nix-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Assured whether relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

### CODE REVIEW

File: pkgs/development/ocaml-modules/opium/default.nix [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+, magic-mime
+, opium_kernel
+, stringext
+
+, alcotest
+}:
+
+buildDunePackage rec {
+	pname = "opium";
+	inherit (opium_kernel) version src meta minimumOCamlVersion;
```
Comment (alexarice) [inline pkgs/development/ocaml-modules/opium/default.nix]: what is the difference between these two packages, one just seems to have more build inputs

File: pkgs/development/ocaml-modules/opium_kernel/default.nix [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+    ppx_sexp_conv ppx_fields_conv
+  ];
+
+	propagatedBuildInputs = [
+    hmap cohttp-lwt ezjsonm
+  ];
+
+  meta = {
+		description = "Sinatra like web toolkit for OCaml based on cohttp & lwt";
+    homepage = https://github.com/rgrinberg/opium;
```
Comment (alexarice) [inline pkgs/development/ocaml-modules/opium_kernel/default.nix]: Quotes round url would be good

File: pkgs/development/ocaml-modules/hmap/default.nix [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+
+  checkPhase = ''
+    runHook preCheck
+    ${topkg.run} build --tests true
+    ${topkg.run} test
+    runHook postCheck
+  '';
+
+  meta = {
+		description = "Hmap provides heterogeneous value maps for OCaml";
```
Comment (alexarice) [inline pkgs/development/ocaml-modules/hmap/default.nix]: Descriptions shouldn't start with the package name

File: pkgs/development/ocaml-modules/hmap/default.nix [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+    homepage = https://erratique.ch/software/hmap;
```
Comment (alexarice) [inline pkgs/development/ocaml-modules/hmap/default.nix]: quotes round url

File: pkgs/development/ocaml-modules/opium/default.nix (hunk 1 above)
Comment (pmahoney) [reply pkgs/development/ocaml-modules/opium/default.nix]: It's just how upstream has organized things. Following examples like [cohttp](https://github.com/NixOS/nixpkgs/blob/master/pkgs/development/ocaml-modules/cohttp/lwt-unix.nix) where a single source is used in multiple derivations.

File: pkgs/development/ocaml-modules/hmap/default.nix [hunk 5]
```diff
@@ -0,0 +1,37 @@
+{ lib
+, buildOcaml
+, fetchurl
+, topkg
+}:
+
+buildOcaml rec {
```
Comment (vbgl) [inline pkgs/development/ocaml-modules/hmap/default.nix]: Please don’t use `buildOcaml`.

File: pkgs/development/ocaml-modules/hmap/default.nix [hunk 6]
```diff
@@ -0,0 +1,37 @@
...
+
+	buildInputs = [ topkg ];
+
+  inherit (topkg) buildPhase installPhase;
+
+  doCheck = true;
+
+  checkPhase = ''
+    runHook preCheck
+    ${topkg.run} build --tests true
```
Comment (vbgl) [inline pkgs/development/ocaml-modules/hmap/default.nix]: IIUC, this line belongs to the `buildPhase`

**Review by vbgl — CHANGES_REQUESTED** [review]

File: pkgs/development/ocaml-modules/hmap/default.nix (hunk 5 above)
Comment (pmahoney) [reply pkgs/development/ocaml-modules/hmap/default.nix]: Is there a deprecation notice somewhere?

File: pkgs/development/ocaml-modules/hmap/default.nix (hunk 6 above)
Comment (pmahoney) [reply pkgs/development/ocaml-modules/hmap/default.nix]: Depends if you want buildPhase building both code and tests (not running tests), or just building code (and leaving the tests to be built in checkPhase). I see other OCaml packages doing what you suggest, so I'll follow that convention.

File: pkgs/development/ocaml-modules/hmap/default.nix (hunk 5 above)
Comment (vbgl) [reply pkgs/development/ocaml-modules/hmap/default.nix]: It is not documented.

**Comment by vbgl** [comment]
@GrahamcOfBorg build ocamlPackages.opium

**Pull request merged**